crates/sc-proof-of-time/src/source/gossip.rs

```diff
@@ -122,70 +195,384 @@ where
         }
     }
 
-    /// Handles the incoming gossip message.
-    async fn handle_incoming_message(&mut self, sender: PeerId, message: GossipCheckpoints) {
-        if let Err(error) = self.incoming_messages_sender.send((sender, message)).await {
-            warn!(%error, "Failed to send incoming message");
+    async fn handle_checkpoints_candidates(
+        &mut self,
+        sender: PeerId,
+        checkpoints: GossipCheckpoints,
+    ) {
+        let next_slot_input = self.state.next_slot_input(atomic::Ordering::Relaxed);
+
+        match checkpoints.slot.cmp(&next_slot_input.slot) {
+            cmp::Ordering::Less => {
+                trace!(
+                    %sender,
+                    slot = %checkpoints.slot,
+                    next_slot = %next_slot_input.slot,
+                    "Checkpoints for outdated slot, ignoring",
+                );
+
+                if let Some(verified_checkpoints) = self
+                    .pot_verifier
+                    .try_get_checkpoints(checkpoints.seed, checkpoints.slot_iterations)
+                {
+                    if verified_checkpoints != checkpoints.checkpoints {
+                        trace!(
+                            %sender,
+                            slot = %checkpoints.slot,
+                            "Invalid old checkpoints, punishing sender",
+                        );
+
+                        self.engine
+                            .lock()
+                            .report(sender, rep::GOSSIP_INVALID_CHECKPOINTS);
+                    }
+                }
+
+                return;
+            }
+            cmp::Ordering::Equal => {
+                if !(checkpoints.seed == next_slot_input.seed
+                    && checkpoints.slot_iterations == next_slot_input.slot_iterations)
+                {
+                    trace!(
+                        %sender,
+                        slot = %checkpoints.slot,
+                        "Checkpoints with next slot mismatch, ignoring",
+                    );
+
+                    self.engine
+                        .lock()
+                        .report(sender, rep::GOSSIP_NEXT_SLOT_MISMATCH);
+                    return;
+                }
+            }
+            cmp::Ordering::Greater => {
+                trace!(
+                    %sender,
+                    slot = %checkpoints.slot,
+                    next_slot = %next_slot_input.slot,
+                    "Checkpoints from the future",
+                );
+
+                self.gossip_cache
+                    .get_or_insert_mut(sender, Default::default)
+                    .push(checkpoints);
+                return;
+            }
+        }
+
+        if self
+            .pot_verifier
+            .verify_checkpoints(
+                checkpoints.seed,
```

Comment:
It's a little confusing that `seed` and `slot_iterations` are part of `checkpoints`

Reply:
I'm open for suggestions on better name here, not perfectly happy with it either

Reply:
I would use `proof` here, proof being a statelessly-verifiable entity that contains inputs (seed and iterations) and outputs (checkpoints). That's what I call it in the spec.

Reply:
`checkpoints: GossipCheckpoints` looks odd, what should data structure be called then, `GossipProofs`?

Reply:
`GossipProof` sounds good to me, as in `Gossip`-ed-`Proof `